File: examples.go
package main

var examples = map[string]string{
	"fib": `# Welcome to the WDTE playground, a browser based evaluation
# environment for WDTE. This playground's features includes the
# standard function set as well as a number of importable modules.
#
# If you have never seen WDTE before and are completely confused at
# the moment, try reading the overview on the WDTE project's wiki:
# https://github.com/DeedleFake/wdte/wiki
#
# For documentation on the standard function set, see
# https://godoc.org/github.com/DeedleFake/wdte/std
#
# The standard library is available for importing, along with the
# following special modules:
#
# * 'canvas' (See the 'Canvas' example.)
# * 'io/file' (This makes no sense in a browser, so it's disabled.)
# * 'io' (Disabled pending https://github.com/gopherjs/gopherjs/issues/705.)
#
# In addition, a print function is provided which uses the Go fmt
# package to create a string representation of its arguments. This
# string is printed to the output pane and then returned.

memo fib n => switch n {
	== 0 => 0;
	== 1 => 1;
	default => + (fib (- n 1)) (fib (- n 2));
};

memo ! n => switch n {
	<= 1 => 1;
	default => - n 1 -> ! -> * n;
};

main =>
	fib 30
	-- print
	-> / 5
	-- print
	;`,

	"stream": `# This example demonstrates the stream module. This module allows
# classic functional iterator operations, such as map, reduce, and
# filter.

'math' => m;
'stream' => s;

main => (
	print 'Map and filter:';
	s.range 0 (* m.pi 2) (/ m.pi 2)
	-> s.map m.sin
	-> s.filter (>= 0)
	-> s.map print
	-> s.drain
	;

	print 'Reduce:';
	s.range 1 5
	-> s.reduce 1 *
	-- print
	;
);`,

	"strings": `# This example demonstrates the strings module. This module provides
# basic string operations, such as finding the index of a substring,
# as well as more complicated operations, such as formatting.

'stream' => s;
'strings' => str;

main => (
	s.new 'abc' 'bcd' 'cde'
	-> s.map (str.index 'cd')
	-> s.collect
	-- print
	;

	'This is the type of English up with which I will not put.'
	-> str.format '{q}'
	-- print
	;
);`,

	"lambda": `# This example demonstrates lambdas by implementing an iterative
# Fibonacci number calculator using streams.

'stream' => s;
'arrays' => a;

fib n => s.range 1 n
	-> s.reduce [0; 1] (@ self p n => [
		a.at p 1;
		+ (a.at p 0) (a.at p 1);
	])
	-> a.at 1
	;

main =>
	fib 30
	-- print
	;`,

	"canvas": `# This example demonstrates the canvas module. This module is a module
# implemented just for this playground. Importing it automatically
# puts the playground in canvas mode, which allows for drawing to the
# output pane. It also redirects the normal output into the error pane.

'math' => m;
'canvas' => c;

# circleRec is the recursive part of circle.
circleRec r cx cy res t p => switch t {
	> (* 2 m.pi) => p -> c.close;
	default => p
             -> c.line (+ cx (* (m.cos t) r)) (+ cy (* (m.sin t) r))
             -> circleRec r cx cy res (+ t (/ m.pi res));
};

# circle returns a circular path centered at (cx, cy) with radius r
# and resolution res.
circle r cx cy res => c.path
                  -> c.move (+ cx r) cy
                  -> circleRec r cx cy res (/ m.pi res);

main => (
	c.start
	-> c.color 'purple'
	-> c.rect 10 10 100 50
	-> c.draw;

	c.start
	-> c.color 'pink'
	-> (
		c.path
		-> c.move 10 50
		-> c.line 30 30
		-> c.line 50 30
		-> c.line 100 100
		-> c.close
	)
	-> c.draw;

	c.start
	-> c.color 'red'
	-> circle 30 100 100 6
	-> c.draw;
);`,
}

File: src/examples.js
// @format

export const fib = {
	name: 'Fibonacci',

	desc: `
Fibonacci
=========

This example provides a memoized implementation of a recursive Fibonacci number generator. It also provides a recursive factorial function for the heck of it.
`,

	input: `let (memo) fib n => n {>= 2 => + (fib (- n 1)) (fib (- n 2))};

let ! n => n {
	<= 1 => 1;
	true => - n 1 -> ! -> * n;
};

fib 30
-- io.writeln io.stdout
-> / 5
-- io.writeln io.stdout
;`,
}

export const fibLike = {
	name: 'Fibonacci-Like Sequence',

	desc: `
Fibonacci-Like Sequence
=======================

This example demonstrates an implementation of a function for determining whether or not a given sequence is similar to a Fibonacci sequence. In other words, it detects if every value in a sequence, starting from the third, is the result of the previous two added together. To accomplish this properly, it defines several extra functions as well:

pop
---

Takes an array and returns a copy that doesn't have the first element of the original.

windows
-------

Takes a size and a stream and returns a new stream that yields a moving window across the orginial stream as arrays of the given size. In other words, if a stream yields 1, then 2, then 3, a stream with windows of size 2 into that stream will yield \`[1; 2]\` and then \`[2; 3]\`

isFibLike
---------

Takes either a stream or an array and returns true if that sequence is Fibonacci-like.`,

	input: `let s => import 'stream';
let a => import 'arrays';

let pop array => a.stream array -> s.skip 1 -> s.collect;

let method windows stream size => s.new
		(stream -> s.limit size -> s.collect)
		(@ next prev =>
			let n => stream -> s.limit 1 -> s.collect;
			len n {
				== 0 => s.end;
				true => a.concat (pop prev) n;
			};
		)
	;

let isFibLike stream => stream {
	reflect "Array" => a.stream stream -> isFibLike;
	true => stream
		-> windows 3
		-> s.all (@ all [p1 p2 n] => == n (+ p1 p2))
		;
	};

[1; 1; 2; 3; 5; 8]
-> isFibLike
-- io.writeln io.stdout
;`,
}

export const stream = {
	name: 'Stream',

	desc: `
Stream
======

This example demonstrates the \`stream\` module. This module provides functional iterator operations, such as map, reduce, and filter.

For a full list of functions, see [the godocs][godoc].

[godoc]: https://www.godoc.org/github.com/DeedleFake/wdte/std/stream
`,

	input: `let m => import 'math';
let s => import 'stream';

io.writeln io.stdout 'Map and filter:';
s.range 0 (* m.pi 2) (/ m.pi 2)
-> s.map m.sin
-> s.filter (>= 0)
-> s.map (io.writeln io.stdout)
-> s.drain
;

io.writeln io.stdout 'Reduce:';
s.range 1 5
-> s.reduce 1 *
-- io.writeln io.stdout
;`,
}

export const strings = {
	name: 'Strings',

	desc: `
Strings
=======

This example demonstrates the \`strings\` module. This module provides basic string operations, such as finding the index of a substring, as well as more complicated operations, such as formatting.

For a full list of functions, including an explanation of the formatting system, see [the godocs][godoc].

[godoc]: https://www.godoc.org/github.com/DeedleFake/wdte/std/strings
`,

	input: `let a => import 'arrays';
let s => import 'stream';
let str => import 'strings';

a.stream ['abc'; 'bcd'; 'cde']
-> s.map (str.index 'cd')
-> s.collect
-- io.writeln io.stdout
;

'This is the type of English up with which I will not put.'
-> str.format '{q}'
-- io.writeln io.stdout
;`,
}

export const lambdas = {
	name: 'Lambdas',

	desc: `
Lambdas
=======

This example demonstrates lambdas by implementing an iterative Fibonacci number calculator using streams.
`,

	input: `let s => import 'stream';
let a => import 'arrays';

let fib n => s.range 1 n
	-> s.reduce [0; 1] (@ self [a b] n => [
		b;
		+ a b;
	])
	-> at 1
	;

fib 30
-- io.writeln io.stdout
;`,
}

export const quine = {
	name: 'Quine',

	desc: `
Quine
=====

This example is an implemenation of a quine. That's about it.
`,

	input: `let str => import 'strings';
let q => "let str => import 'strings';\\nlet q => {q};\\nstr.format q q -- io.writeln io.stdout;";
str.format q q -- io.writeln io.stdout;`,
}

export const hundredDoors = {
	name: '100 Doors',

	desc: `
100 Doors
=========

The [100 doors problem](https://www.rosettacode.org/wiki/100_doors), as presented by Rosetta Code, is as follows:

There are 100 doors that are all closed. You walk past the doors in the same direction 100 times. On the first pass, you toggle the state of every door, opening closed doors and closing open doors. On the second pass, you toggle every second door. On the third you toggle every third door. Etc.

This example simulates this scenario, printing out the final state of the doors.
`,

	input: `let a => import 'arrays';
let s => import 'stream';
let str => import 'strings';

let toggle doors m =>
	a.stream doors
	-> s.enumerate
	-> s.map (@ s [i v] => [+ i 1; v])
	-> s.map (@ s [i v] => % i m {
			== 0 => ! v;
			true => v;
		})
	-> s.collect
	;

s.range 100
-> s.map false
-> s.collect : doors
-> s.range 1 100
-> s.reduce doors toggle
-> a.stream
-> s.enumerate
-> s.map (@ s [i v] =>
		0 {
			v => 'Open';
			true => 'Closed';
		}
		-> str.format '{}: {}' (+ i 1)
		-- io.writeln io.stdout;
	)
-> s.drain
;`,
}

//export const Canvas = {
//	desc: ``,
//	input: ``,
//}
